pennylane/templates/subroutines/all_singles_doubles.py

```diff
@@ -0,0 +1,181 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the AllSinglesDoubles template.
+"""
+# pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
+import numpy as np
+import pennylane as qml
+from pennylane.operation import Operation, AnyWires
+from pennylane.ops import BasisState
+
+
+class AllSinglesDoubles(Operation):
+    r"""Builds a quantum circuit to prepare correlated states of molecules
+    by applying all :class:`~.pennylane.SingleExcitation` and
+    :class:`~.pennylane.DoubleExcitation` operations associated with all
+    single and double excitations of the initial Hartree-Fock state.
+
+    The template initializes the :math:`n`-qubit system to encode
+    the input Hartree-Fock state and applies the particle-conserving
+    :class:`~.pennylane.SingleExcitation` and
+    :class:`~.pennylane.DoubleExcitation` operations which are implemented as
+    `Givens rotations <https://en.wikipedia.org/wiki/Givens_rotation>`_ that act
+    on the subspace of two and four qubits, respectively. The total number of
+    excitation gates and the indices of the qubits they act on are obtained
+    using the :func:`~.excitations` function.
+
+    For example, the quantum circuit for the case of two electrons and six qubits
+    is sketched in the figure below,
+
+    |
+
+    .. figure:: ../../_static/templates/subroutines/all_singles_doubles.png
+        :align: center
+        :width: 70%
+        :target: javascript:void(0);
+
+    |
+
+    In this case, we have four single and double excitations that preserve the total-spin
+    projection of the Hartree-Fock state. The :class:`~.pennylane.SingleExcitation` gate
+    :math:`G` act on the qubits ``[0, 2], [0, 4], [1, 3], [1, 5]`` as indicated by the
+    squares, while the :class:`~.pennylane.DoubleExcitation` operation :math:`G^{(2)}` is
+    applied to the qubits ``[0, 1, 2, 3], [0, 1, 2, 5], [0, 1, 2, 4], [0, 1, 4, 5]``.
+
+    The resulting unitary conserves the number of particles and prepares the
+    :math:`n`-qubit system in a superposition of the initial Hartree-Fock state and
+    other states encoding multiply-excited configurations.
+
+    Args:
+        weights (tensor_like): size ``(len(singles) + len(doubles),)`` tensor containing the
+            angles entering the :class:`~.pennylane.SingleExcitation` and
+            :class:`~.pennylane.DoubleExcitation` operations
```

Comment:
```suggestion
            :class:`~.pennylane.DoubleExcitation` operations, in that order
```
So user knows the first weights are for single excitations followed by the double excitation ones